### PR TITLE
[Impeller] Enable fixed-rate compression support in Vulkan.

### PR DESCRIPTION
**Pull request by chinmaygarde**
Fixes https://github.com/flutter/flutter/issues/129501

### CODE REVIEW

**Comment by chinmaygarde** [comment]
Followup of https://github.com/flutter/engine/pull/53079 with fixups (see the second patch) after guidance from Greg & Jesse.

I don't have a device on hand that has these extensions to check if this actually works and I don't think the swapchain stuff is expected to work on the Pixels (I'll still wire it up). So still WIP. Will finish verification next week when I get back to my desk.

**Comment by jonahwilliams** [comment]
I can confirm that my pixel 8 is selecting `vk::ImageCompressionFixedRateFlagBitsEXT::e4Bpc`! That said, I don't know if its doing anything :)

**Comment by chinmaygarde** [comment]
Got a hold of a Pixel 8a. The diffs aren't noticeable (but are present).
AFRC

**Comment by jonahwilliams** [comment]
interesting, those are all raster images. Is there a diff on a page with a lot of text?

**Comment by chinmaygarde** [comment]
Curiously, on a wall of text, I could find no differences. Thinking I messed up, I tried 4 different builds with logging. Again, no difference. Thinking that that AFRC conditions were not met, I tried logging like so but didn't see any logs either. So whatever its doing, there seems to be no difference in rendered output. I almost can't believe it.

```c++
if (!image_info_chain.isLinked<vk::ImageCompressionControlEXT>() &&
    desc.compression_type == CompressionType::kLossy) {
  FML_LOG(ERROR)
      << "Requested lossy compression but could not satisfy conditions.";
}
```

Text

**Comment by chinmaygarde** [comment]
I'm going to back out the swapchain stuff because that is not meant to work. And what we have here we are doing on iOS too. So we have some experience with this. I'm going to get memory usage numbers first though.

**Comment by jonahwilliams** [comment]
is the AFRC only applying to offscreen rendering and not the onscreen ?

**Comment by chinmaygarde** [comment]
Yes, but the images themselves should have AFRC enabled.

**Comment by chinmaygarde** [comment]
Rapidly swiping the Wonders in the Wondrous app shows the memory usage drops from ~730 MB to ~520 MB reliable. I tried different sections for similar results. I think this matches up with what we were observing on iOS too right?

Screenshot 2024-06-17 at 3 02 23 PM
Screenshot 2024-06-17 at 2 56 57 PM

**Comment by jonahwilliams** [comment]
Yes that seems about right for ballpark. We saw about 50% for iOS, this is less than that but I don't know what else is in that heap.

**Comment by chinmaygarde** [comment]
Ok, I've cleanup up the swapchain stuff and verified similar behavior to iOS with lossy compression. Though this took longer than expected, I'm satisfied we are doing the best we can and there is parity between the platforms. We can try to wire up the onscreen surfaces compression or mess with other rates at a later time. But I'd like to close this loop on this one. PTAL.

File: impeller/renderer/backend/vulkan/capabilities_vk.cc [hunk 1]
```diff
@@ -611,4 +641,63 @@ bool CapabilitiesVK::HasExtension(OptionalDeviceExtensionVK ext) const {
...
+}
+
+std::optional<vk::ImageCompressionFixedRateFlagBitsEXT>
+CapabilitiesVK::GetSupportedFRCRate(CompressionType compression_type,
+                                    const FRCFormatDescriptor& desc) const {
+  if (compression_type != CompressionType::kLossy) {
+    return std::nullopt;
+  }
+  if (!supports_texture_fixed_rate_compression_) {
+    return std::nullopt;
```
Comment (jonahwilliams) [inline impeller/renderer/backend/vulkan/capabilities_vk.cc:655]: nit: combine these into one?

```
if (compression_type != CompressionType::kLossy || !supports_texture_fixed_rate_compression_) {
  return std::nullopt;
}
```

**Review by jonahwilliams — COMMENTED** [review]
I'm not sure why this is crashing on the tests...

**Comment by chinmaygarde** [comment]
The failures look real. I probably forgot to unlink an item in the structure chain. Taking a look.

File: impeller/renderer/backend/vulkan/capabilities_vk.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/renderer/backend/vulkan/capabilities_vk.cc:666]: Done.

**Comment by chinmaygarde** [comment]
I believe I have fixed everything and addressed all comments. PTAL.

**Comment by chinmaygarde** [comment]
Ping @jonahwilliams. I've addressed all comments verified memory savings. A review please?

**Review by jonahwilliams — APPROVED** [review]
LGTM!

**Pull request merged**